src/pf/existing/update_source_files.cpp

```diff
@@ -0,0 +1,192 @@
+                                                  std::make_reverse_iterator(begin_fn),
+                                                  [](char c) { return std::isblank(c); });
+    if (*last_source_list_char == '\n') {
+        ++last_source_list_char;
+    }
+
+    if (last_source_list_char.base() <= insertion_point) {
+        // If the source list is empty, we need to ensure there's a newline separating it
+        // from the rest
+        return cmakelists.insert(insertion_point, '\n');
+    }
+    return cmakelists.erase(insertion_point, last_source_list_char.base());
+}
+
+auto insert_sources(std::string&                    cmakelists,
+                    std::string::iterator           insertion_point,
+                    std::vector<std::string> const& sources,
+                    std::string const&              indent) {
+    bool first_iteration = true;
+
+    for (auto const& source : sources) {
+        if (!first_iteration) {
+            insertion_point = std::next(cmakelists.insert(insertion_point, '\n'));
+        } else {
+            first_iteration = false;
+        }
+
+        insertion_point
+            = std::next(cmakelists.insert(insertion_point, indent.begin(), indent.end()),
+                        indent.size());
+        insertion_point
+            = std::next(cmakelists.insert(insertion_point, source.begin(), source.end()),
+                        source.size());
+    }
+
+    return insertion_point;
+}
+
+std::pair<std::string::iterator, std::string::iterator>
+write_sources(std::string&                    cmakelists,
+              std::vector<std::string> const& source_strings,
+              std::string::iterator           insertion_comment,
+              std::string::iterator           begin_fn,
+              std::string::iterator           end_fn) {
+    std::string const indent
+        = ::get_indent_at_indicator_comment(insertion_comment, begin_fn, end_fn);
+    auto insertion_point = ::find_insertion_location(insertion_comment, begin_fn, end_fn);
+
+    // Note: invalidates other iterators
+    insertion_point = ::erase_existing_sources(cmakelists, insertion_point, begin_fn, end_fn);
+
+    auto const end_insertion
+        = ::insert_sources(cmakelists, insertion_point, source_strings, indent);
+    return std::pair{end_insertion, cmakelists.end()};
+}
+
+}  // namespace
+
+void pf::update_source_files(fs::path const&              project_root,
+                             std::vector<fs::path> const& source_files) {
+    fs::path const src_dir        = project_root / "src";
+    fs::path const src_cmakelists = src_dir / "CMakeLists.txt";
+
+    if (fs::exists(src_cmakelists)) {
+        std::vector<std::string> const source_strings
+            = ::relative_source_strings(source_files, src_dir);
+
+        std::string cmakelists = [&] {
+            std::ostringstream out;
+            out << std::ifstream{src_cmakelists}.rdbuf();
```

Comment:
Us `pf::open()`. Maybe a `pf::read_file(fs::path)` would be useful? I could imagine "Read the entire file" would be a useful and common operation.